### PR TITLE
Implement parallel downloads

### PR DESCRIPTION
**Pull request by dimaryaz**
We lose precise progress bar, though: we don't know the object sizes, so can only count the number of objects.

### CODE REVIEW

**Comment by asah** [comment]
@dimaryaz interesting... appveyor error looks legit - take a look?

**Comment by dimaryaz** [comment]
Oops, yes, broken test. Will fix.

File: compiler/quilt/tools/command.py [hunk 1]
```diff
@@ -879,84 +887,125 @@ def install(package, hash=None, version=None, tag=None, force=False):
 
     pkgobj = store.install_package(owner, pkg, response_contents)
 
-    with requests.Session() as s3_session:
-        total = len(response_urls)
-        for idx, (download_hash, url) in enumerate(sorted(iteritems(response_urls))):
-            print("Downloading %s (%d/%d)..." % (download_hash, idx + 1, total))
+    obj_queue = sorted(iteritems(response_urls), reverse=True)
```
Comment (kevinemoore) [inline compiler/quilt/tools/command.py]: Curious, why reverse?

File: compiler/quilt/tools/command.py [hunk 2]
```diff
@@ -879,84 +887,125 @@ def install(package, hash=None, version=None, tag=None, force=False):
...
-            if file_hash != download_hash:
-                os.remove(temp_path)
-                raise CommandException("Fragment hashes do not match: expected %s, got %s." %
-                                       (download_hash, file_hash))
+                    if not success:
+                        # We've already printed an error, so not much to do - just move on to the next object.
+                        continue
+
+                    # Ungzip the downloaded fragment.
+                    temp_path = store.temporary_object_path(obj_hash)
```
Comment (kevinemoore) [inline compiler/quilt/tools/command.py]: Maybe some of this code can move into PackageStore?

File: compiler/quilt/tools/command.py (hunk 1 above)
Comment (dimaryaz) [reply compiler/quilt/tools/command.py]: `obj_queue` is actually a stack (just cause I can use a normal list for that) - so I reverse it so it shows up sorted when popping items off the end.

File: compiler/quilt/tools/command.py (hunk 2 above)
Comment (dimaryaz) [reply compiler/quilt/tools/command.py]: Which part of it? The ungzipping code?

**Comment by dimaryaz** [comment]
Yeah, it's definitely getting worse - though it's hard to refactor cause there are so many variables involved. (Like, I'd have a helper function that takes 10 params, etc.)

It's particularly bad with nested loops - and I had to add an extra variable just to break out of the inner loop, then continue the outer loop. But anyways, I'll try to do that in a follow-up PR.

**Pull request merged**